lib/wallaroo/core/sink/connector_sink/connector_sink_notify.pony

```diff
@@ -0,0 +1,455 @@
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ implied. See the License for the specific language governing
+ permissions and limitations under the License.
+
+*/
+
+use "backpressure"
+use "buffered"
+use "net"
+use "wallaroo/core/common"
+use "wallaroo/core/network"
+use "wallaroo_labs/bytes"
+use cp = "wallaroo_labs/connector_protocol"
+use "wallaroo_labs/mort"
+
+class ConnectorSinkNotify
+  var _fsm_state: cp.ConnectorProtoFsmState = cp.ConnectorProtoFsmDisconnected
+  var _header: Bool = true
+  var _connected: Bool = false
+  var _throttled: Bool = false
+  let _stream_id: cp.StreamId = 1
+  let _sink_id: RoutingId
+  let _worker_name: WorkerName
+  let _protocol_version: String
+  let _cookie: String
+  let _auth: ApplyReleaseBackpressureAuth
+  let stream_name: String
+  var credits: U32 = 0
+  var acked_point_of_ref: cp.MessageId = 0
+  var message_id: cp.MessageId = acked_point_of_ref
+  var _connection_count: USize = 0
+  // 2PC
+  var _rtag: U64 = 77777
+  var twopc_intro_done: Bool = false
+  var twopc_txn_id_last_committed: (None|String) = None
+  var twopc_uncommitted_list: (None|Array[String] val) = None
+  let twopc_reconnect_buffer: Array[(String val | Array[U8] val)] = twopc_reconnect_buffer.create()
+
+  new create(sink_id: RoutingId, worker_name: WorkerName,
+    protocol_version: String, cookie: String,
+    auth: ApplyReleaseBackpressureAuth)
+  =>
+    _sink_id = sink_id
+    _worker_name = worker_name
+    _protocol_version = protocol_version
+    _cookie = cookie
+    _auth = auth
+
+    stream_name = "worker-" + worker_name + "-id-" + _sink_id.string()
+
+  fun ref accepted(conn: WallarooOutgoingNetworkActor ref) =>
+    Unreachable()
+
+  fun ref auth_failed(conn: WallarooOutgoingNetworkActor ref) =>
+    Unreachable()
+
+  fun ref connecting(conn: WallarooOutgoingNetworkActor ref, count: U32) =>
+    None
+
+  fun ref connected(conn: WallarooOutgoingNetworkActor ref) =>
+    @printf[I32]("ConnectorSink connected\n".cstring())
+    _header = true
+    _connected = true
+    _throttled = false
+    twopc_intro_done = false
+    twopc_uncommitted_list = None
+    _connection_count = _connection_count + 1
+    // Apply runtime throttle until we're done with initial 2PC ballet.
+    throttled(conn)
+    conn.expect(4)
+
+    // TODO: configure connector v2 program string
+    // TODO: configure connector v2 instance_name string
+    let hello = cp.HelloMsg(_protocol_version, _cookie,
+      "a program", "an instance")
+    send_msg(conn, hello)
+
+    // 2PC: We don't know how many transactions the sink has that
+    // have been waiting for a phase 2 message.  We need to discover
+    // their txn_id strings and abort them.
```

Comment:
We can only abort them if the checkpoint was aborted.  This sounds like we are simply aborting anything that has not yet been committed.

Reply:
updated